src/VisualQueryBuilder/components/OperationEditor.tsx

```diff
@@ -0,0 +1,336 @@
+import { css, cx } from '@emotion/css';
+import React, { useEffect, useState } from 'react';
+import { Draggable } from 'react-beautiful-dnd';
+import { v4 } from 'uuid';
+
+import { DataSourceApi, GrafanaTheme2, TimeRange } from '@grafana/data';
+import { Button, Icon, InlineField, Tooltip, useTheme2 } from '@grafana/ui';
+
+import { OperationHeader } from './OperationHeader';
+import {
+  QueryBuilderOperation,
+  QueryBuilderOperationDefinition,
+  QueryBuilderOperationParamDef,
+  QueryBuilderOperationParamValue,
+  VisualQueryModeller,
+} from '../types';
+import { Stack } from '../../QueryEditor/Stack';
+import { getOperationParamEditor, getOperationParamId } from './OperationParamEditor';
+
+interface Props {
+  operation: QueryBuilderOperation;
+  index: number;
+  query: any;
+  datasource: DataSourceApi;
+  queryModeller: VisualQueryModeller;
+  onChange: (index: number, update: QueryBuilderOperation) => void;
+  onRemove: (index: number) => void;
+  onRunQuery: () => void;
+  flash?: boolean;
+  highlight?: boolean;
+  timeRange?: TimeRange;
+  isConflictingOperation?: (operation: QueryBuilderOperation, otherOperations: QueryBuilderOperation[]) => boolean;
+  innerQueryPlaceholder: string;
+}
+
+export function OperationEditor({
+  operation,
+  index,
+  onRemove,
+  onChange,
+  onRunQuery,
+  queryModeller,
+  query,
+  datasource,
+  flash,
+  highlight,
+  timeRange,
+  isConflictingOperation,
+  innerQueryPlaceholder,
+}: Props) {
+  const def = queryModeller.getOperationDefinition(operation.id);
+  const shouldFlash = useFlash(flash);
+  const id = v4();
+
+  const theme = useTheme2();
+  const isConflicting = isConflictingOperation ? isConflictingOperation(operation, query.operations) : false;
+  const styles = getStyles(theme, isConflicting);
+
+  if (!def) {
+    return <span>Operation {operation.id} not found</span>;
+  }
+
+  const onParamValueChanged = (paramIdx: number, value: QueryBuilderOperationParamValue) => {
+    const update: QueryBuilderOperation = { ...operation, params: [...operation.params] };
+    update.params[paramIdx] = value;
+    callParamChangedThenOnChange(def, update, index, paramIdx, onChange);
+  };
+
+  const onAddRestParam = () => {
+    const update: QueryBuilderOperation = { ...operation, params: [...operation.params, ''] };
+    callParamChangedThenOnChange(def, update, index, operation.params.length, onChange);
+  };
+
+  const onRemoveRestParam = (paramIdx: number) => {
+    const update: QueryBuilderOperation = {
+      ...operation,
+      params: [...operation.params.slice(0, paramIdx), ...operation.params.slice(paramIdx + 1)],
+    };
+    callParamChangedThenOnChange(def, update, index, paramIdx, onChange);
+  };
+
+  const operationElements: React.ReactNode[] = [];
+
+  for (let paramIndex = 0; paramIndex < operation.params.length; paramIndex++) {
```

Comment:
To add as well: this could be extracted to another component and memoized. Otherwise all these components are regenerated on every render.

Reply:
Issue created: https://github.com/grafana/grafana-experimental/issues/102